server/src/metrics/metric_batching.rs

```diff
@@ -0,0 +1,328 @@
+}
+
+impl SizedItemKind {
+    pub fn serialized_len(&self) -> usize {
+        match self {
+            SizedItemKind::App(app) => size_of_batch(app),
+            SizedItemKind::Metric(metric) => size_of_batch(metric),
+            SizedItemKind::Impact(impact) => size_of_batch(impact),
+        }
+    }
+}
+
+impl SizedItem {
+    pub fn new(item: SizedItemKind) -> SizedItem {
+        SizedItem {
+            size: item.serialized_len(),
+            item,
+        }
+    }
+}
+struct ByteCounter {
+    count: usize,
+}
+
+impl Write for ByteCounter {
+    fn write(&mut self, buf: &[u8]) -> io::Result<usize> {
+        self.count += buf.len();
+        Ok(buf.len())
+    }
+
+    fn flush(&mut self) -> io::Result<()> {
+        Ok(())
+    }
+}
+
+fn partition_batch(mut batch: MetricsBatch, max_batch_size: usize) -> Vec<MetricsBatch> {
+    let mut sized_items = Vec::new();
+
+    // applications are necessarily first, since Unleash needs to know the application in
+    // order to store the metrics. Impact metrics are last by coincidence since they're independent
+    sized_items.extend(
+        batch
+            .applications
+            .drain(..)
+            .map(|app| SizedItem::new(SizedItemKind::App(app))),
+    );
+    sized_items.extend(
+        batch
+            .metrics
+            .drain(..)
+            .map(|metric| SizedItem::new(SizedItemKind::Metric(metric))),
+    );
+    sized_items.extend(
+        batch
+            .impact_metrics
+            .drain(..)
+            .map(|impact| SizedItem::new(SizedItemKind::Impact(impact))),
+    );
+
+    let mut batches = Vec::new();
+    let mut current_batch = MetricsBatch::default();
+    let mut current_size = EMPTY_BATCH_JSON_OVERHEAD;
+
+    for item in sized_items {
+        let next_size = current_size + item.size + 1; // for commas separating items
+
+        if next_size > max_batch_size {
+            if !current_batch.applications.is_empty()
+                || !current_batch.metrics.is_empty()
+                || !current_batch.impact_metrics.is_empty()
+            {
+                batches.push(current_batch);
+            }
+            current_batch = MetricsBatch::default();
+            current_size = EMPTY_BATCH_JSON_OVERHEAD;
+        }
+
+        match item.item {
+            SizedItemKind::App(app) => current_batch.applications.push(app),
+            SizedItemKind::Metric(metric) => current_batch.metrics.push(metric),
+            SizedItemKind::Impact(impact) => current_batch.impact_metrics.push(impact),
+        }
+
+        current_size += item.size + 1; // for commas separating items
+    }
+
+    if !current_batch.applications.is_empty()
+        || !current_batch.metrics.is_empty()
+        || !current_batch.impact_metrics.is_empty()
+    {
+        batches.push(current_batch);
+    }
+
+    batches
+}
+
+#[instrument(skip(batch))]
+pub(crate) fn cut_into_sendable_batches(batch: MetricsBatch) -> Vec<MetricsBatch> {
```

Comment:
This is the exposed function that got redesigned. This is a thin wrapper around `partition_batch` so that we can prop test the internals against a much smaller max bound - the prop tests are expensive and a smaller bound means faster execution